hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -353,8 +386,77 @@ private void processMaintenanceOnlyIndexes(
         DatanodeDetails target = iterator.next();
         commands.put(target, replicateCommand);
         additionalMaintenanceCopiesNeeded -= 1;
+        maintIndexes.remove(replica.getReplicaIndex());
+        createdIndexes.add(replica.getReplicaIndex());
+      }
+    }
+  }
+
+  /**
+   * Function processes Replicas to fix placement policy issues.
+   * @param replicas
+   * @param deletionInFlight
+   * @param container
+   * @param excludedNodes
+   * @param sources
+   * @param createdIndexes
+   * @param placementStatus
+   * @throws IOException
+   */
+  private Map<DatanodeDetails, SCMCommand<?>> processMisreplication(
+          Set<ContainerReplica> replicas,
+          List<DatanodeDetails> deletionInFlight, ContainerInfo container,
+          List<DatanodeDetails> excludedNodes,
+          Map<Integer, Pair<ContainerReplica, NodeStatus>> sources,
+          Set<Integer> createdIndexes, ContainerPlacementStatus placementStatus)
+          throws IOException {
+    if (placementStatus.isPolicySatisfied() ||
+            placementStatus.misReplicationCount() <= createdIndexes.size()) {
+      return Collections.emptyMap();
+    }
+    Map<DatanodeDetails, SCMCommand<?>> commands = new HashMap<>();
+    List<Integer> sortedReplicaIdx =
```

Comment:
I'm not sure what this is doing? Are we sorting the replicas so that if we have say, 2 copies of index 3, it will be first in the list?

To fix mis-replication, we must have some indexes that reside on the same rack. So if we find a new node on a new rack, a candidate index to move is one where there are multiple indexes on the same rack. So I think we need to find any indexes that reside on the same rack and then move one of them to the new target.

Note there could be 2 or more on the same rack.

Reply:
We shouldn't handle rack count in the ReplicationManager, we have offloaded this requirement on the placement policy where we can choose nodes based on the used nodes & asking for additional replicas

Reply:
The placementPolicy gives us the new target, but I don't understand what this code is doing to select which replica index it is going to copy onto that new target.

Say we have EC 3-2, and the current layout looks like:

```
index: 1 Rack: 1
Index:2 Rack: 2 **
Index:3 Rack: 2 **
Index:4 Rack: 4
Index:5 Rack: 5
```
When we goto the placement policy, it says "here is a new target on rack 3".

That means we need to make a new copy of index 2 or 3 onto rack 3, and then later remove the original index 2 or 3, but that is handled later by the over rep handler. Copying index 1, 4 or 5 to Rack 3 does not improve the placement, as we are still short by a rack.

How does the logic here address that, as you seem to be sorting the DNs in order of duplicate index count first, and then using that below to pick which node to make a new copy of.